File: src/MetabolicSimulator.Mobile/src/engine/bioMath.ts
// bioMath.ts - Core biochemical kinetics

export class BioMath {
  /**
   * Calculates reaction rate using Michaelis-Menten kinetics: V = (Vmax * [S]) / (Km + [S])
   */
  static michaelisMenten(vmax: number, km: number, substrateConc: number): number {
    if (substrateConc <= 0) return 0;
    return (vmax * substrateConc) / (km + substrateConc);
  }

  /**
   * Handles competitive inhibition: V = Vmax * [S] / (Km * (1 + [I]/Ki) + [S])
   */
  static competitiveInhibition(vmax: number, km: number, s: number, i: number, ki: number): number {
    if (s <= 0) return 0;
    const kmApp = km * (1.0 + (i / ki));
    return (vmax * s) / (kmApp + s);
  }

  /**
   * Handles allosteric activation (Simplified Hill-like): V = Vmax * [S] / (Km + [S]) * (1 + [A]/Ka)
   */
  static activation(baseRate: number, activatorConc: number, ka: number): number {
    // Simple multiplier logic for activation
    // If activator >> Ka, rate increases.
    // For Michaelis-Menten, often modeled as Vmax_app = Vmax * (1 + A/Ka)
    return baseRate * (1.0 + (activatorConc / ka));
  }
}

File: src/MetabolicSimulator.Domain/Core/BioMath.cs
namespace MetabolicSimulator.Domain.Core;

public static class BioMath
{
    /// <summary>
    /// Calculates reaction rate using Michaelis-Menten kinetics: V = (Vmax * [S]) / (Km + [S])
    /// </summary>
    public static double MichaelisMenten(double vmax, double km, double substrateConc)
    {
        if (substrateConc <= 0) return 0;
        return (vmax * substrateConc) / (km + substrateConc);
    }

    /// <summary>
    /// Handles competitive inhibition: V = Vmax * [S] / (Km * (1 + [I]/Ki) + [S])
    /// </summary>
    public static double CompetitiveInhibition(double vmax, double km, double s, double i, double ki)
    {
        if (s <= 0) return 0;
        var kmApp = km * (1.0 + (i / ki));
        return (vmax * s) / (kmApp + s);
    }

    /// <summary>
    /// Handles non-competitive inhibition: V = (Vmax / (1 + [I]/Ki)) * [S] / (Km + [S])
    /// </summary>
    public static double NonCompetitiveInhibition(double vmax, double km, double s, double i, double ki)
    {
        if (s <= 0) return 0;
        var vmaxApp = vmax / (1.0 + (i / ki));
        return (vmaxApp * s) / (km + s);
    }
}
